validator/client/src/main/java/tech/pegasys/teku/validator/client/restapi/apis/PostVoluntaryExit.java

```diff
@@ -58,17 +84,79 @@ public PostVoluntaryExit() {
             .response(SC_OK, "Success response", SIGNED_VOLUNTARY_EXIT_RESPONSE_TYPE)
             .withNotFoundResponse()
             .withUnauthorizedResponse()
             .withForbiddenResponse()
-            .withNotImplementedResponse()
             .build());
+    this.spec = spec;
+    this.keyManager = keyManager;
+    this.validatorApiChannel = validatorApiChannel;
+    this.timeProvider = timeProvider;
   }
 
   @Override
   public void handleRequest(RestApiRequest request) throws JsonProcessingException {
-    request.respondError(SC_NOT_IMPLEMENTED, "Not implemented");
+    final BLSPublicKey publicKey = request.getPathParameter(PARAM_PUBKEY_TYPE);
+    final UInt64 epoch =
+        request
+            .getOptionalQueryParameter(EPOCH_PARAMETER)
+            .orElseGet(
+                () -> {
+                  final SpecVersion genesisSpec = spec.getGenesisSpec();
+                  final UInt64 genesisTime = genesisSpec.getConfig().getMinGenesisTime();
+                  final UInt64 currentTime = timeProvider.getTimeInSeconds();
+                  final UInt64 slot =
+                      genesisSpec.miscHelpers().computeSlotAtTime(genesisTime, currentTime);
+                  return spec.computeEpochAtSlot(slot);
+                });
+
+    final SafeFuture<SignedVoluntaryExit> future = createSignedVoluntaryExit(publicKey, epoch);
+    request.respondAsync(future.thenApply(AsyncApiResponse::respondOk));
+  }
+
+  private SafeFuture<SignedVoluntaryExit> createSignedVoluntaryExit(
+      final BLSPublicKey publicKey, final UInt64 epoch) {
+    return validatorApiChannel
+        .getGenesisData()
+        .thenCompose(
+            genesisData -> {
+              if (genesisData.isEmpty()) {
+                throw new InvalidConfigurationException(
+                    "Unable to fetch genesis data, cannot generate an exit.");
+              }
+
+              final Bytes32 genesisRoot = genesisData.get().getGenesisValidatorsRoot();
+              final Fork fork = spec.getForkSchedule().getFork(epoch);
+              final ForkInfo forkInfo = new ForkInfo(fork, genesisRoot);
+
+              return validatorApiChannel
+                  .getValidatorIndices(Set.of(publicKey))
+                  .thenApply(
+                      indicesMap -> {
+                        final int validatorIndex = indicesMap.get(publicKey);
+                        final Validator validator =
+                            keyManager.getActiveValidatorKeys().stream()
+                                .filter(v -> v.getPublicKey().equals(publicKey))
+                                .findFirst()
+                                .orElseThrow(
+                                    () ->
+                                        new BadRequestException(
+                                            "Validator not found with public key value."));
+                        final VoluntaryExit message =
```

Comment:
maybe something like "Validator {} is not in the list of keys managed by this service."